File: platform/sdk/swift/Tests/LayerXSDKTests/NativeCapabilitySetTests.swift
import Foundation
import XCTest
@testable import LayerXSDK

final class NativeCapabilitySetTests: XCTestCase {
    private func field(_ object: [String: Any], _ name: String) throws -> Data {
        let text = try XCTUnwrap(object[name] as? String)
        XCTAssertEqual(text.utf8.count % 2, 0)
        var result = Data(); var offset = text.startIndex
        while offset < text.endIndex {
            let end = text.index(offset, offsetBy: 2)
            result.append(try XCTUnwrap(UInt8(text[offset..<end], radix: 16))); offset = end
        }
        return result
    }
    private func amount(_ object: [String: Any]) throws -> UInt128Value {
        let text = try XCTUnwrap(object["maximum_amount"] as? String)
        _ = try ProtocolAmount(text)
        var encoded = [UInt8](repeating: 0, count: 16)
        for digit in text.utf8 {
            var carry = Int(digit - 48)
            for index in encoded.indices.reversed() {
                let product = Int(encoded[index]) * 10 + carry
                encoded[index] = UInt8(product & 255); carry = product >> 8
            }
            XCTAssertEqual(carry, 0)
        }
        return .init(high: encoded.prefix(8).reduce(0) { ($0 << 8) | UInt64($1) }, low: encoded.suffix(8).reduce(0) { ($0 << 8) | UInt64($1) })
    }
    private func logicalGrants(_ values: [[String: Any]]) throws -> [NativeCapability] {
        try values.map { value in
            let grant: NativeCapability
            switch try XCTUnwrap(value["kind"] as? String) {
            case "StorageRead": grant = .storageRead
            case "StorageWrite": grant = .storageWrite
            case "EmitEvent": grant = .emitEvent
            case "Call": grant = .call(program: try field(value, "program"))
            case "Transfer402": grant = .transfer402(asset: try field(value, "asset"), to: try field(value, "to"), maximumAmount: try amount(value))
            case "ProgramSpend": grant = .programSpend(ownerProgram: try field(value, "owner_program"), seed: try field(value, "seed"), sourceAccount: try field(value, "source_account"), asset: try field(value, "asset"), to: try field(value, "to"), maximumAmount: try amount(value))
            case "ReceiptRead": grant = .receiptRead(receiptDigest: try field(value, "receipt_digest"))
            case "BalanceView": grant = .balanceView(account: try field(value, "account"), asset: try field(value, "asset"), receiptDigest: try field(value, "receipt_digest"))
            case "SharedStorageRead": grant = .sharedStorageRead
            case "SharedStorageWrite": grant = .sharedStorageWrite
            default: throw NativeProgramCallError.invalid
            }
            XCTAssertEqual(Int(try NativeCapabilitySet.encode([grant])[2]), try XCTUnwrap(value["tag"] as? Int))
            return grant
        }
    }
    private func assertEncoding(_ grants: [NativeCapability], _ expected: Data) throws {
        XCTAssertEqual(try NativeCapabilitySet.encode(grants), expected)
        XCTAssertEqual(try NativeCapabilitySet.encode(NativeCapabilitySet.decode(expected)), expected)
    }
    func testRuntimeFixtureBindsLogicalGrantsAndEscalations() throws {
        var path = URL(fileURLWithPath: #filePath)
        for _ in 0..<6 { path.deleteLastPathComponent() }
        path.appendPathComponent("platform/sdk/conformance/fixtures/native-program-capabilities-v2.json")
        let fixture = try XCTUnwrap(JSONSerialization.jsonObject(with: Data(contentsOf: path)) as? [String: Any])
        let values = try XCTUnwrap(fixture["capabilities"] as? [[String: Any]])
        XCTAssertEqual(try values.map { try XCTUnwrap($0["tag"] as? Int) }, [1, 2, 3, 4, 5, 9, 6, 10, 7, 8])
        let parent = try logicalGrants(values)
        try assertEncoding(parent, field(fixture, "canonical_hex"))
        let requested = try logicalGrants(XCTUnwrap(fixture["narrowed_capabilities"] as? [[String: Any]]))
        try assertEncoding(requested, field(fixture, "narrowed_hex"))
        let narrowed = try NativeCapabilitySet.narrow(parent, to: requested)
        try assertEncoding(narrowed, field(fixture, "narrowed_hex"))
        XCTAssertEqual(fixture["equal_narrowing_accepted"] as? Bool, true)
        try assertEncoding(NativeCapabilitySet.narrow(parent, to: parent), field(fixture, "canonical_hex"))
        let escalations = try XCTUnwrap(fixture["escalation_cases"] as? [[String: Any]])
        XCTAssertEqual(escalations.count, 3)
        for escalation in escalations {
            XCTAssertEqual(escalation["parent"] as? String, "narrowed")
            XCTAssertEqual(escalation["accepted"] as? Bool, false)
            let grants = try logicalGrants(XCTUnwrap(escalation["capabilities"] as? [[String: Any]]))
            try assertEncoding(grants, field(escalation, "canonical_hex"))
            XCTAssertThrowsError(try NativeCapabilitySet.narrow(narrowed, to: grants))
        }
    }

    private func identifier(_ value: UInt8) -> Data { Data(repeating: value, count: 32) }

    func testBoundsAndNarrowingMatchRuntime() throws {
        let owner = identifier(1), asset = identifier(2), to = identifier(3), receipt = identifier(4)
        let seed = Data([0, 255]); let source = try NativeCapabilitySet.deriveProgramAccount(owner: owner, seed: seed)
        let maximum = UInt128Value(high: 1, low: 0)
        let spend = NativeCapability.programSpend(ownerProgram: owner, seed: seed, sourceAccount: source, asset: asset, to: to, maximumAmount: maximum)
        let view = NativeCapability.balanceView(account: source, asset: asset, receiptDigest: receipt)
        let parent: [NativeCapability] = [.sharedStorageWrite, .sharedStorageRead, view, .receiptRead(receiptDigest: receipt), spend,
            .transfer402(asset: asset, to: to, maximumAmount: maximum), .call(program: owner), .emitEvent, .storageWrite, .storageRead]
        let encoded = try NativeCapabilitySet.encode(parent)
        XCTAssertEqual(encoded.prefix(5), Data([0, 10, 1, 2, 3]))
        let reduced = NativeCapability.programSpend(ownerProgram: owner, seed: seed, sourceAccount: source, asset: asset, to: to, maximumAmount: .init(high: 0, low: UInt64.max))
        XCTAssertEqual(try NativeCapabilitySet.narrow(parent, to: [reduced, view]).count, 2)
        let escalated = NativeCapability.programSpend(ownerProgram: owner, seed: seed, sourceAccount: source, asset: asset, to: to, maximumAmount: .init(high: 1, low: 1))
        XCTAssertThrowsError(try NativeCapabilitySet.narrow(parent, to: [escalated]))
        let changedView = NativeCapability.balanceView(account: source, asset: asset, receiptDigest: identifier(5))
        XCTAssertThrowsError(try NativeCapabilitySet.narrow(parent, to: [changedView]))
        XCTAssertThrowsError(try NativeCapabilitySet.encode([view, changedView]))
        XCTAssertThrowsError(try NativeCapabilitySet.narrow([], to: [.storageRead]))
        XCTAssertThrowsError(try NativeCapabilitySet.encode([.transfer402(asset: asset, to: to, maximumAmount: .init(high: 0, low: 0))]))
        XCTAssertThrowsError(try NativeCapabilitySet.encode([.call(program: Data(repeating: 0, count: 32))]))
        XCTAssertThrowsError(try NativeCapabilitySet.encode([.programSpend(ownerProgram: owner, seed: seed, sourceAccount: identifier(9), asset: asset, to: to, maximumAmount: maximum)]))
        XCTAssertThrowsError(try NativeCapabilitySet.encode([.programSpend(ownerProgram: owner, seed: Data(repeating: 0, count: 129), sourceAccount: source, asset: asset, to: to, maximumAmount: maximum)]))
        for length in 0..<encoded.count { XCTAssertThrowsError(try NativeCapabilitySet.decode(encoded.prefix(length))) }
        XCTAssertThrowsError(try NativeCapabilitySet.decode(encoded + Data([0])))
        for malformed: [UInt8] in [[0, 2, 2, 1], [0, 2, 1, 1], [0, 1, 11], [0, 239]] {
            XCTAssertThrowsError(try NativeCapabilitySet.decode(Data(malformed)))
        }
        let views: [NativeCapability] = (1...33).map { .balanceView(account: identifier(UInt8($0)), asset: asset, receiptDigest: receipt) }
        _ = try NativeCapabilitySet.encode(Array(views.prefix(32)))
        XCTAssertThrowsError(try NativeCapabilitySet.encode(views))
        var full: [NativeCapability] = []
        for index in 0...NativeCapabilitySet.maximumGrants {
            var fullSeed = Data(repeating: 0, count: 128); fullSeed[0] = UInt8(index >> 8); fullSeed[1] = UInt8(truncatingIfNeeded: index)
            let account = try NativeCapabilitySet.deriveProgramAccount(owner: owner, seed: fullSeed)
            full.append(.programSpend(ownerProgram: owner, seed: fullSeed, sourceAccount: account, asset: asset, to: to, maximumAmount: maximum))
        }
        XCTAssertEqual(try NativeCapabilitySet.encode(Array(full.prefix(NativeCapabilitySet.maximumGrants))).count, NativeCapabilitySet.maximumBytes)
        XCTAssertThrowsError(try NativeCapabilitySet.encode(full))
    }
}

File: platform/sdk/dotnet/tests/LayerX.Sdk.Tests/NativeCapabilitySetTests.cs
using System.Numerics;
using System.Runtime.CompilerServices;
using System.Text.Json;
using LayerX.Sdk;
using Xunit;
using static LayerX.Sdk.NativeCapabilitySet;

namespace LayerX.Sdk.Tests;

public sealed class NativeCapabilitySetTests
{
    private static byte[] Field(JsonElement value, string name) => Convert.FromHexString(value.GetProperty(name).GetString()!);
    private static List<Capability> LogicalGrants(JsonElement values)
    {
        var grants = new List<Capability>();
        foreach (var value in values.EnumerateArray())
        {
            Capability grant = value.GetProperty("kind").GetString() switch
            {
                "StorageRead" => new StorageRead(),
                "StorageWrite" => new StorageWrite(),
                "EmitEvent" => new EmitEvent(),
                "Call" => new Call(Field(value, "program")),
                "Transfer402" => new Transfer402(Field(value, "asset"), Field(value, "to"), BigInteger.Parse(value.GetProperty("maximum_amount").GetString()!, System.Globalization.CultureInfo.InvariantCulture)),
                "ProgramSpend" => new ProgramSpend(Field(value, "owner_program"), Field(value, "seed"), Field(value, "source_account"), Field(value, "asset"), Field(value, "to"), BigInteger.Parse(value.GetProperty("maximum_amount").GetString()!, System.Globalization.CultureInfo.InvariantCulture)),
                "ReceiptRead" => new ReceiptRead(Field(value, "receipt_digest")),
                "BalanceView" => new BalanceView(Field(value, "account"), Field(value, "asset"), Field(value, "receipt_digest")),
                "SharedStorageRead" => new SharedStorageRead(),
                "SharedStorageWrite" => new SharedStorageWrite(),
                _ => throw new ArgumentException("Unknown fixture capability"),
            };
            Assert.Equal(value.GetProperty("tag").GetByte(), Encode([grant])[2]); grants.Add(grant);
        }
        return grants;
    }
    private static void AssertEncoding(IReadOnlyList<Capability> grants, byte[] expected)
    {
        Assert.Equal(expected, Encode(grants)); Assert.Equal(expected, Encode(Decode(expected)));
    }
    private static string FixturePath([CallerFilePath] string path = "")
    {
        var root = Path.GetDirectoryName(path)!;
        for (var depth = 0; depth < 5; depth++) root = Path.GetDirectoryName(root)!;
        return Path.Combine(root, "platform/sdk/conformance/fixtures/native-program-capabilities-v2.json");
    }
    [Fact]
    public void RuntimeFixtureBindsLogicalGrantsAndEscalations()
    {
        using var document = JsonDocument.Parse(File.ReadAllText(FixturePath())); var fixture = document.RootElement;
        Assert.Equal(new byte[] { 1, 2, 3, 4, 5, 9, 6, 10, 7, 8 }, fixture.GetProperty("capabilities").EnumerateArray().Select(value => value.GetProperty("tag").GetByte()).ToArray());
        var parent = LogicalGrants(fixture.GetProperty("capabilities")); AssertEncoding(parent, Field(fixture, "canonical_hex"));
        var requested = LogicalGrants(fixture.GetProperty("narrowed_capabilities")); AssertEncoding(requested, Field(fixture, "narrowed_hex"));
        var narrowed = Narrow(parent, requested); AssertEncoding(narrowed, Field(fixture, "narrowed_hex"));
        Assert.True(fixture.GetProperty("equal_narrowing_accepted").GetBoolean()); AssertEncoding(Narrow(parent, parent), Field(fixture, "canonical_hex"));
        Assert.Equal(3, fixture.GetProperty("escalation_cases").GetArrayLength());
        foreach (var escalation in fixture.GetProperty("escalation_cases").EnumerateArray())
        {
            Assert.Equal("narrowed", escalation.GetProperty("parent").GetString()); Assert.False(escalation.GetProperty("accepted").GetBoolean());
            var grants = LogicalGrants(escalation.GetProperty("capabilities")); AssertEncoding(grants, Field(escalation, "canonical_hex"));
            Assert.Throws<ArgumentException>(() => Narrow(narrowed, grants));
        }
    }

    private static byte[] Identifier(byte value) => Enumerable.Repeat(value, 32).ToArray();
    [Fact]
    public void BoundsAndNarrowingMatchRuntime()
    {
        var owner = Identifier(1); var asset = Identifier(2); var to = Identifier(3); var receipt = Identifier(4);
        byte[] seed = [0, 255]; var source = DeriveProgramAccount(owner, seed); var maximum = BigInteger.One << 64;
        var spend = new ProgramSpend(owner, seed, source, asset, to, maximum); var view = new BalanceView(source, asset, receipt);
        Capability[] parent = [new SharedStorageWrite(),
            new SharedStorageRead(),
            view,
            new ReceiptRead(receipt),
            spend,
            new Transfer402(asset, to, maximum),
            new Call(owner),
            new EmitEvent(),
            new StorageWrite(),
            new StorageRead()];
        var encoded = Encode(parent); Assert.Equal(new byte[] { 0, 10, 1, 2, 3 }, encoded[..5]);
        Assert.Equal(2, Narrow(parent, [spend with { MaximumAmount = maximum - 1 }, view]).Count);
        Assert.Throws<ArgumentException>(() => Narrow(parent, [spend with { MaximumAmount = maximum + 1 }]));
        var changedView = view with { ReceiptDigest = Identifier(5) };
        Assert.Throws<ArgumentException>(() => Narrow(parent, [changedView]));
        Assert.Throws<ArgumentException>(() => Encode([view, changedView]));
        Assert.Throws<ArgumentException>(() => Narrow([], [new StorageRead()]));
        Assert.Throws<ArgumentException>(() => Encode([new Transfer402(asset, to, 0)]));
        Assert.Throws<ArgumentException>(() => Encode([new Transfer402(asset, to, BigInteger.One << 128)]));
        Assert.Throws<ArgumentException>(() => Encode([new Call(new byte[32])]));
        Assert.Throws<ArgumentException>(() => Encode([spend with { SourceAccount = Identifier(9) }]));
        Assert.Throws<ArgumentException>(() => Encode([spend with { Seed = new byte[129] }]));
        for (var length = 0; length < encoded.Length; length++) Assert.Throws<ArgumentException>(() => Decode(encoded[..length]));
        Assert.Throws<ArgumentException>(() => Decode([.. encoded, 0]));
        foreach (var malformed in new byte[][] { [0, 2, 2, 1], [0, 2, 1, 1], [0, 1, 11], [0, 239] })
            Assert.Throws<ArgumentException>(() => Decode(malformed));
        var views = Enumerable.Range(1, 33).Select(index => (Capability)new BalanceView(Identifier((byte)index), asset, receipt)).ToArray();
        _ = Encode(views[..32]); Assert.Throws<ArgumentException>(() => Encode(views));
        var full = new List<Capability>();
        for (var index = 0; index <= MaximumGrants; index++)
        {
            var fullSeed = new byte[128]; fullSeed[0] = (byte)(index >> 8); fullSeed[1] = (byte)index;
            full.Add(new ProgramSpend(owner, fullSeed, DeriveProgramAccount(owner, fullSeed), asset, to, maximum));
        }
        Assert.Equal(MaximumBytes, Encode(full.GetRange(0, MaximumGrants)).Length);
        Assert.Throws<ArgumentException>(() => Encode(full));
    }
}
